### PR TITLE
[0.73] Fix new role prop after JS-Shim was removed by Meta

### PR DESCRIPTION
**Pull request by FalseLobster**
## Summary:

Prior to 0.73, the new `role` prop was remapped to `accessibilityRole` on the JS-side.  Starting with #37304, the work needs to be completed on the native side.  Since the new prop is ARIA inspired, the mappings are taking from the [ARIA Core AAM](https://www.w3.org/TR/core-aam-1.2/#mapping_role_table) which disagrees with some of the mappings used in the old `accessibilityRole` prop.  Users of the old prop are unaffected, but the new prop will take the mappings from the spec.

## Test Plan:

Tested a variety of permutations of accessibilityRole and role to confirm the behavior looks correct in Accessibility Inspector

### CODE REVIEW

File: packages/rn-tester/RNTesterPods.xcodeproj/project.pbxproj [hunk 1]
```diff
@@ -1460,8 +1460,6 @@
 				OTHER_LDFLAGS = (
 					"-ObjC",
 					"-lc++",
-					"-Wl",
-					"-ld_classic",
```
Comment (Saadnajmi) [inline packages/rn-tester/RNTesterPods.xcodeproj/project.pbxproj:1464]: I think we can undo these changes and the podfile.lock change. These flags are added by the react native cocoapods install script for some Xcode 15 compatibility (because I think for a while, Xcode 15 used a new linker that broke react native). Not sure if it's still needed so I've just avoided committing it.

File: packages/react-native/React/Base/RCTConvert.m [hunk 2]
```diff
@@ -1487,6 +1487,105 @@ + (NSPropertyList)NSPropertyList:(id)json
     integerValue)
 
 #if TARGET_OS_OSX // [macOS
+// Used to convert the JSON from `role` to the corresponding
+// NSAccessibilityRole (NSString*), has slightly different mappings than the
```
Comment (Saadnajmi) [inline packages/react-native/React/Base/RCTConvert.m:1491]: I think we don't need to mention `Used to convert the JSON from `role` to the corresponding
// NSAccessibilityRole (NSString*)` since that's the purpose of this whole file. The rest about "slightly different mappings" seems more important to have first

File: packages/react-native/React/Base/RCTConvert.m [hunk 3]
```diff
@@ -1487,6 +1487,105 @@ + (NSPropertyList)NSPropertyList:(id)json
...
+// old accessibilityRole prop for back compatability. `role` matches ARIA Core
+// AAM spec and takes precedence. See https://www.w3.org/TR/core-aam-1.1/
++ (NSString*) accessibilityRoleFromRole:(NSString*)ariaRole
```
Comment (amgleitman) [inline packages/react-native/React/Base/RCTConvert.m:1494]: The argument is an `NSString *` here but an `id` in the header. I didn't even know this was legal in Objective C.

File: packages/react-native/React/Base/RCTConvert.m [hunk 4]
```diff
@@ -1487,6 +1487,105 @@ + (NSPropertyList)NSPropertyList:(id)json
...
+      @"none": NSAccessibilityGroupRole,
+      @"note": NSAccessibilityGroupRole,
+      @"option": NSAccessibilityStaticTextRole,
+      @"presentation": NSAccessibilityGroupRole,
+      @"progressbar": NSAccessibilityProgressIndicatorRole,
+      @"radio": NSAccessibilityRadioButtonRole,
+      @"radiogroup": NSAccessibilityRadioGroupRole,
+      @"region": NSAccessibilityGroupRole,
+      @"row": NSAccessibilityRowRole,
+      // rowgroup not mapped
```
Comment (Saadnajmi) [inline packages/react-native/React/Base/RCTConvert.m:1549]: Is there like an unknown role or is it better to leave omitted?

File: packages/react-native/React/Base/RCTConvert.m [hunk 5]
```diff
@@ -1487,6 +1487,105 @@ + (NSPropertyList)NSPropertyList:(id)json
...
+// Used to convert the JSON from `role` to the corresponding
+// NSAccessibilityRole (NSString*), has slightly different mappings than the
+// old accessibilityRole prop for back compatability. `role` matches ARIA Core
+// AAM spec and takes precedence. See https://www.w3.org/TR/core-aam-1.1/
++ (NSString*) accessibilityRoleFromRole:(NSString*)ariaRole
+{
+  static NSDictionary<NSString *, NSAccessibilityRole> * ariaRoleToNsRole;
+  static dispatch_once_t onceToken;
+  dispatch_once(&onceToken, ^{
+    ariaRoleToNsRole = @{
```
Comment (Saadnajmi) [inline packages/react-native/React/Base/RCTConvert.m:1499]: nit: prefer `ariaRoleToNSAccessibilityRole`

File: packages/react-native/React/Base/RCTConvert.h [hunk 6]
```diff
@@ -154,7 +154,8 @@ typedef BOOL css_backface_visibility_t;
 + (RCTTextDecorationLineType)RCTTextDecorationLineType:(id)json;
 
 #if TARGET_OS_OSX // [macOS
-+ (NSString *)accessibilityRoleFromTraits:(id)json;
++ (NSString *)accessibilityRoleFromTraits:(id)json withAriaMappings:(BOOL) useAriaMappings;
++ (NSString *)accessibilityRoleFromRole:(id)json;
```
Comment (amgleitman) [inline packages/react-native/React/Base/RCTConvert.h:158]: Perhaps rename to `accessibilityRoleFromAriaRole:` to make it clearer?

File: packages/react-native/React/Base/RCTConvert.m [hunk 7]
```diff
@@ -1487,6 +1487,105 @@ + (NSPropertyList)NSPropertyList:(id)json
...
+      @"textbox": NSAccessibilityTextFieldRole,
+      @"timer": NSAccessibilityGroupRole,
+      @"toolbar": NSAccessibilityToolbarRole,
+      @"tooltip": NSAccessibilityGroupRole,
+      @"tree": NSAccessibilityOutlineRole,
+      @"treegrid": NSAccessibilityTableRole,
+      @"treeitem": NSAccessibilityRowRole,
+    };
+  });
+  NSAccessibilityRole nsRole = [ariaRoleToNsRole valueForKey: ariaRole];
```
Comment (Saadnajmi) [inline packages/react-native/React/Base/RCTConvert.m:1573]: I think this might crash if the key doesn't exist (one of the annoying things about NSDictionaries) so you may want to check the key exists first. There should be examples elsewhere

File: packages/react-native/React/Base/RCTConvert.m [hunk 8]
```diff
@@ -1538,13 +1637,13 @@ + (NSString*)accessibilityRoleFromTrait:(NSString*)trait
   return role;
 }
 
-+ (NSString *)accessibilityRoleFromTraits:(id)json
++ (NSString *)accessibilityRoleFromTraits:(id)json withAriaMappings:(BOOL) useAriaMappings
```
Comment (amgleitman) [inline packages/react-native/React/Base/RCTConvert.m:1640]: nit: remove extra space before `useAriaMappings`, as well as before `json` in the two invocations to `accessibilityRoleFromRole:` below

File: packages/react-native/React/Base/RCTConvert.m [hunk 9]
```diff
@@ -1487,6 +1487,105 @@ + (NSPropertyList)NSPropertyList:(id)json
...
+      @"treeitem": NSAccessibilityRowRole,
+    };
+  });
+  NSAccessibilityRole nsRole = [ariaRoleToNsRole valueForKey: ariaRole];
+  if (nsRole == nil) {
+    // Dictionary above is fully filled for roles spec'd by aria mappings since
+    // the aria mappings disagrees with some of the old mappings. We'll still 
+    // fall back to the legacy mappings for macOS specific roles like 
+    // `disclosure` and legacy accessibilityTrait-based mappings like 
+    // `adjustable`.
```
Comment (Saadnajmi) [inline packages/react-native/React/Base/RCTConvert.m:1579]: Maybe something more succinct like "Fall back to legacy mappings if an aria mapping is not found. This would include macOS specific roles like `disclosure` and legacy accessibilityTrait-based mappings like `adjustable`.". Also for comments longer than 3 lines, I think I prefer `/* .... */

File: packages/react-native/React/Base/RCTConvert.m [hunk 10]
```diff
@@ -1487,6 +1487,105 @@ + (NSPropertyList)NSPropertyList:(id)json
...
+    nsRole = [RCTConvert accessibilityRoleFromTrait:ariaRole];
+  }
+  return nsRole;
+}
+
+// Used to convert the JSON from accessibilityRole to the corresponding
+// NSAccessibilityRole (NSString*), has slightly different mappings than the
+// new role prop for back compatability. role matches ARIA spec and takes
+// precedence.
```
Comment (Saadnajmi) [inline packages/react-native/React/Base/RCTConvert.m:1588]: Similar comment about mentioning JSON

File: packages/react-native/React/Base/RCTConvert.m (hunk 8 above)
Comment (Saadnajmi) [inline packages/react-native/React/Base/RCTConvert.m:1640]: Nit:  I think the more objective-c ish way to write this would be `usingAriaMappings:(BOOK) useAriaMappings`. "withAriaMappings" feels like an object is about to be passed in to me

Comment (Saadnajmi) [reply packages/react-native/React/Base/RCTConvert.m:1640]: But also our official guidelines are on `aka.ms/objc-guide`
https://microsoft.github.io/objc-guide/MethodsAndImplementations/Naming.html

File: packages/react-native/React/Views/RCTViewManager.m [hunk 11]
```diff
@@ -292,34 +291,45 @@ - (RCTShadowView *)shadowView
...
-  #else // [macOS
-    if (json) {
-      view.reactAccessibilityElement.accessibilityRole = [RCTConvert accessibilityRoleFromTraits:json];
-    } else {
-      view.reactAccessibilityElement.accessibilityRole = defaultView.accessibilityRole;
-    }
-  #endif // macOS]
+#else // [macOS
+  // accessibilityRoleInternal is used to cache the converted value from the prop
+  view.reactAccessibilityElement.accessibilityRoleInternal = json ? [RCTConvert accessibilityRoleFromTraits:json withAriaMappings:NO] : nil;
```
Comment (Saadnajmi) [inline packages/react-native/React/Views/RCTViewManager.m:304]: Should we cache in a local variable instead?

File: packages/react-native/React/Views/RCTViewManager.m [hunk 12]
```diff
@@ -292,34 +291,45 @@ - (RCTShadowView *)shadowView
...
+#if !TARGET_OS_OSX // [macOS]
   UIAccessibilityTraits roleTraits = json ? [RCTConvert UIAccessibilityTraits:json] : UIAccessibilityTraitNone;
   if (view.reactAccessibilityElement.roleTraits != roleTraits) {
     view.roleTraits = roleTraits;
     view.reactAccessibilityElement.role = json ? [RCTConvert NSString:json] : nil;
     [self updateAccessibilityTraitsForRole:view withDefaultView:defaultView];
   }
+#else // [macOS
+  // role is used to cache the converted value from the prop
+  view.reactAccessibilityElement.role = json ? [RCTConvert accessibilityRoleFromTraits:json withAriaMappings:YES] : nil;
```
Comment (Saadnajmi) [inline packages/react-native/React/Views/RCTViewManager.m:324]: similar comment about caching

**Comment by Saadnajmi** [comment]
An aside, we'll want to eventually also implement for Fabric, but not blocking for 0.73

File: packages/react-native/React/Base/RCTConvert.m (hunk 7 above)
Comment (FalseLobster) [reply packages/react-native/React/Base/RCTConvert.m:1573]: It does not crash. If the key doesn't exist, nsRole is nil and the `if` block is entered. I'm assuming this is ok?

File: packages/react-native/React/Base/RCTConvert.m (hunk 4 above)
Comment (FalseLobster) [reply packages/react-native/React/Base/RCTConvert.m:1549]: I debated this too. The ARIA spec is a little vague about what "Not Mapped" means in this instance.  The current implementation is going to fail to find the key in the dictionary and fall back to `NSAccessibilityUnknownRole`.  Arguably, I think we might be better off using the defaultView's accessibilityRole in that situation.  Do you have an opinion?

File: packages/react-native/React/Base/RCTConvert.m (hunk 8 above)
Comment (FalseLobster) [reply packages/react-native/React/Base/RCTConvert.m:1640]: Thanks, I'm still getting used to the style conventions in Objective C

**Comment by Saadnajmi** [comment]
Q: Would it make sense for our `main` branch version of the PR to use the aria mappings for the accessibilityRole slot, rather than the legacy mappings we defined? I I can imagine we keep the legacy mappings for one version (0.73) and disable it in the next (0.74). Alternatively, maybe put it behind a Feature flag via this [JS file ](https://github.com/microsoft/react-native-macos/blob/main/packages/react-native/Libraries/ReactNative/ReactNativeFeatureFlags.js)or this [native file](https://github.com/microsoft/react-native-macos/blob/70e4f2fe703d28c6d47f739e83914309fad0d492/packages/react-native/React/Base/RCTConstants.m#L37) ?

File: packages/react-native/React/Base/RCTConvert.m (hunk 8 above)
Comment (FalseLobster) [reply packages/react-native/React/Base/RCTConvert.m:1640]: Thanks!

File: packages/react-native/React/Views/RCTViewManager.m (hunk 11 above)
Comment (FalseLobster) [reply packages/react-native/React/Views/RCTViewManager.m:304]: I'm not entirely sure what you mean. Local to this function? Caching ViewManager wouldn't make sense, right? That's one-to-many (RCTViewManager to UIViews). We cache on the actual UIView so that regardless of the order of operations between setting `accessibilityRole` and `role` `updateAccessibilityRole` has access to both converted values and can correctly apply the precedence we want. I know it's kind of an edge case for users to be setting/unsetting both props, but that's how Meta implemented and spec'd the behavior of these props.

Prior to this change, accessibilityRoleInternal and role already existed but were never actually used.  This updates them to be used in a fashion that parallels the iOS implementation Meta does.

**Comment by FalseLobster** [comment]
> Q: Would it make sense for our `main` branch version of the PR to use the aria mappings for the accessibilityRole slot, rather than the legacy mappings we defined? I I can imagine we keep the legacy mappings for one version (0.73) and disable it in the next (0.74). Alternatively, maybe put it behind a Feature flag via this [JS file ](https://github.com/microsoft/react-native-macos/blob/main/packages/react-native/Libraries/ReactNative/ReactNativeFeatureFlags.js)or this [native file](https://github.com/microsoft/react-native-macos/blob/70e4f2fe703d28c6d47f739e83914309fad0d492/packages/react-native/React/Base/RCTConstants.m#L37) ?

Maybe, I'm open to it. I kinda suspect at some point Meta will drop accessibilityRole like they did accessibilityTraits (even though our impl still has accessibilityTraits).  I think we run the risk internally of breaking partners who are using accessibilityRole if we silently change some mappings for them, whereas only applying the new mappings to the new values to the new prop feels safe, which is worth considering, but I suppose that's why you're calling out the feature flag :)

File: packages/react-native/React/Views/RCTViewManager.m (hunk 11 above)
Comment (Saadnajmi) [reply packages/react-native/React/Views/RCTViewManager.m]: Ah, ok, we're handling the edge case of when things are set in a different order. Fabric makes it easier because views have a reference to all their props, and you just setall the props from the property bag and then do any "final" logic after.

File: packages/react-native/React/Base/RCTConvert.h [hunk 13]
```diff
@@ -154,7 +154,7 @@ typedef BOOL css_backface_visibility_t;
 + (RCTTextDecorationLineType)RCTTextDecorationLineType:(id)json;
 
 #if TARGET_OS_OSX // [macOS
-+ (NSString *)accessibilityRoleFromTraits:(id)json;
++ (NSString *)accessibilityRoleFromTraits:(id)json usingAriaMappings:(BOOL) useAriaMappings;
```
Comment (Saadnajmi) [inline packages/react-native/React/Base/RCTConvert.h:157]: Sorry, I'm gonna be dumb. I think the style is `useAriaMappings:(BOOL)useAriaMappings`. Redundant, yes, hence why Swift makes it nicer. Sorry for the double change :/

File: packages/react-native/React/Base/RCTConvert.m (hunk 4 above)
Comment (Saadnajmi) [reply packages/react-native/React/Base/RCTConvert.m:1548]: If by `defaultView` you mean the one that argument that shows up in RCTConvert / view manager code, I don't think that's ever used. If you mean like the default accessibilityRole of an NSView, what is that? I thought it might be nil and then VoiceOver would just skip over it?

File: packages/react-native/React/Views/RCTViewManager.m [hunk 14]
```diff
@@ -292,34 +291,45 @@ - (RCTShadowView *)shadowView
...
+  view.reactAccessibilityElement.accessibilityRoleInternal = json ? [RCTConvert accessibilityRoleFromTraits:json usingAriaMappings:NO] : nil;
+  // update the actual NSAccessibilityRole if it doesn't match
+  if (view.reactAccessibilityElement.accessibilityRole != view.reactAccessibilityElement.accessibilityRoleInternal) {
+    [self updateAccessibilityRole:view withDefaultView:defaultView];
+  }
+#endif // macOS]
 }
 
-#if !TARGET_OS_OSX // [macOS]
+
```
Comment (Saadnajmi) [inline packages/react-native/React/Views/RCTViewManager.m:312]: Nit: Do we want the extra space? Asking because during merges, I have spent time removing random extra spaces we've introduced 😅

File: packages/react-native/React/Views/RCTViewManager.m [hunk 15]
```diff
@@ -336,7 +346,17 @@ - (void)updateAccessibilityTraitsForRole:(RCTView *)view withDefaultView:(RCTVie
...
 }
-#endif // [macOS]
+#else // [macOS
+- (void) updateAccessibilityRole:(RCTView *)view withDefaultView:(RCTView *)defaultView
+{
+  // First check the value from `role`
+  view.reactAccessibilityElement.accessibilityRole = view.reactAccessibilityElement.role ? view.reactAccessibilityElement.role : nil;
+  // Fallback to `accessibilityRole` if nil, or the defaultView's NSAccessibilityRole
+  if (view.reactAccessibilityElement.accessibilityRole == nil) {
+    view.reactAccessibilityElement.accessibilityRole = view.reactAccessibilityElement.accessibilityRoleInternal ? view.reactAccessibilityElement.accessibilityRoleInternal : defaultView.accessibilityRole;
```
Comment (Saadnajmi) [inline packages/react-native/React/Views/RCTViewManager.m:356]: OK, you mean this defaultView. TBH, this feels like unreachable code and the equivalent of setting to nil. I guess my preference is to copy whatever iOS does? And/or check if Fabric does something different ( I think the file is `packages/react-native/ReactCommon/react/renderer/components/view/accessibilityPropsConversions.h`)

Comment (FalseLobster) [reply packages/react-native/React/Views/RCTViewManager.m:356]: iOS maps this to UIAccessibilityTraitNone... so the question is should we map to NSAccessibilityRoleUnknown? nil? Group?  All of those have slightly different implications on the accessibility representation, I would guess.

Comment (FalseLobster) [reply packages/react-native/React/Views/RCTViewManager.m:355]: I updated it to nil in the latest iteration, which I think may be the most accurate representation?

Comment (Saadnajmi) [reply packages/react-native/React/Views/RCTViewManager.m:355]: Yeah I looked around Fabric and saw that iOS has `none` while macOS doesn't.. I think nil is probably most accurate.

File: packages/react-native/React/Base/RCTConvert.m [hunk 16]
```diff
@@ -1487,6 +1487,104 @@ + (NSPropertyList)NSPropertyList:(id)json
...
 #if TARGET_OS_OSX // [macOS
+// This is for the role prop & has slightly different mappings than the
+// old accessibilityRole prop for back compatability. `role` matches ARIA Core
+// AAM spec and takes precedence. See https://www.w3.org/TR/core-aam-1.1/
++ (NSString*) accessibilityRoleFromAriaRole:(NSString*)ariaRole
+{
+  // rowgroup is explicitly not mapped
+  if ([ariaRole isEqualToString:@"rowgroup"]) {
+      return nil;
+  }
```
Comment (Saadnajmi) [inline packages/react-native/React/Base/RCTConvert.m:1498]: Rip, but also, I don't even know if I know what a row group is :D

**Review by Saadnajmi — APPROVED** [review]
